### PR TITLE
Jc/openapi3

### PR DESCRIPTION
**Pull request by joeconwaystk**
This PR accomplishes two things:

1. Significantly improves the router's capabilities and performance by using a tree search instead of just iterating over routes. This also helps recognize route collisions prior to starting the server, which will now reveal themselves at startup.

2. Introduces a much improved documentation generating scheme for the OpenAPI 3.0 specification. While not fully complete, the structure is there and the majority of the items in the specification are automatically generated. There are a handful of issues in pivotal for fully completing this implementation, but do require some improvements to the auth implementation first.

### CODE REVIEW

**Comment by anachlas** [comment]
👍

File: lib/base/documentable.dart [hunk 1]
```diff
@@ -1,189 +1,419 @@
...
-    new APISecurityItem()
-      ..name = "token"
-      ..description = "Authorization: Bearer {Token}"
-      ..type = APISecurityType.oauth2
-      ..tokenURL = ""
-      ..flow = "password"
-  ];
+class APIContact {
+  String name;
+  String URL;
```
Comment (smswz) [inline lib/base/documentable.dart]: should we make this lower case to not confuse with a type?

**Comment by joeconwaystk** [comment]
done

File: lib/base/documentable.dart [hunk 2]
```diff
@@ -1,189 +1,419 @@
...
-    if (securityItems != null) {
-      var sec = {};
-      securityItems.forEach((si) {
-        sec[si.name] = si.asMap();
-      });
-      m["securityDefinitions"] = sec;
-    }
+class APILicense {
+  String name;
+  String URL;
```
Comment (smswz) [inline lib/base/documentable.dart]: lowercase URL

File: lib/base/request_path.dart [hunk 3]
```diff
@@ -39,222 +39,325 @@ class RequestPath {
...
+      } else {
+        buffer.writeCharCode(code);
+        insideExpression = true;
+      }
+    } else if (code == closeExpression) {
+      if (insideExpression) {
+        buffer.writeCharCode(code);
+        insideExpression = false;
+      } else {
+        buffer.writeCharCode(code);
```
Comment (smswz) [inline lib/base/request_path.dart]: should this not throw like above? since you shouldn't have a `)` except for bounding the expression?

Comment (joeconwaystk) [inline lib/base/request_path.dart]: yeah, that makes sense

File: lib/base/request_path.dart [hunk 4]
```diff
@@ -54,17 +54,64 @@ class RoutePathSpecification implements APIDocumentable {
...
+
+      if (!seg.matches(iter.current)) {
+        return false;
+      }
+    }
+
+    return iter.moveNext() == false;
+  }
+
+  RequestPath requestPathForSegments(List<String> requestSegments) {
```
Comment (smswz) [inline lib/base/request_path.dart]: why not use a constructor?

Comment (joeconwaystk) [inline lib/base/request_path.dart]: agreed, will change

File: lib/base/documentable.dart [hunk 5]
```diff
@@ -260,7 +261,7 @@ class APIParameter {
     m["description"] = description;
     m["required"] = (parameterLocation == APIParameterLocation.path ? true : required);
     m["deprecated"] = deprecated;
-    m["schema"] = schemaObject.asMap();
+    m["schema"] = schemaObject?.asMap();
```
Comment (smswz) [inline lib/base/documentable.dart]: Why is the type called `APISchemaObject` vs `APISchema` and likewise the variable name?

File: test/base/document_test.dart [hunk 6]
```diff
@@ -13,7 +13,10 @@ main() {
     expect(resolvedPath.startsWith("$homeDir/.pub-cache/hosted/pub.dartlang.org"), true);
   });
 
-  test("Documentation test", () {
+  test("App-to-router test", () {
+    var app = new Application<TPipeline>();
+    var doc = app.document(new PackagePathResolver(new File(".packages").path));
+    print("${doc.asMap()}");
```
Comment (smswz) [inline test/base/document_test.dart]: This doesnt test anything

File: lib/base/documentable.dart (hunk 5 above)
Comment (joeconwaystk) [inline lib/base/documentable.dart]: APISchema seemed too broad reaching for its purpose. Technically, all of these API*'s are called 'Objects' in the specification, but obviously that part gets dropped for most.

File: test/base/document_test.dart (hunk 6 above)
Comment (joeconwaystk) [inline test/base/document_test.dart]: That's been there for a bit, I'm not sure if there is a reasonable way to test documentation, but I will remove this test.

File: lib/base/request_path.dart [hunk 7]
```diff
@@ -136,13 +136,23 @@ class RoutePathSpecification implements APIDocumentable {
...
+
           if (pathParamNames.length != opPathParamNames.length) {
             return false;
           }
+
           return pathParamNames.every((p) => opPathParamNames.contains(p.name));
         }).toList();
 
+    // Strip operation parameters that are already in the path, but move their type into
+    // the path's path parameters
```
Comment (smswz) [inline lib/base/request_path.dart]: path's path parameters?

File: test/base/document_test.dart [hunk 8]
```diff
@@ -48,11 +48,11 @@ class TController extends HTTPController {
   /// ABCD
```
Comment (smswz) [inline test/base/document_test.dart]: what?

Comment (joeconwaystk) [inline test/base/document_test.dart]: This is a documentation comment, so the documentation parser actually reads this information for the description of a HTTP operation

File: lib/base/request_path.dart [hunk 9]
```diff
@@ -98,13 +98,13 @@ class RoutePathSpecification implements APIDocumentable {
     return p;
   }
 
-  /// Returns an [APIPath].
+  /// Returns a [List] one [APIPath].
```
Comment (smswz) [inline lib/base/request_path.dart]: what?

File: lib/auth/authentication_server.dart [hunk 10]
```diff
@@ -167,7 +167,7 @@ class AuthenticationServer<ResourceOwner extends Authenticatable, TokenType exte
     };
   }
 
-  /// A utility method to generate a password hash using the PBKDF2 scheme.
+  /// A utility method to generate am password hash using the PBKDF2 scheme.
```
Comment (smswz) [inline lib/auth/authentication_server.dart]: typo

File: test/base/document_test.dart (hunk 8 above)
Comment (smswz) [inline test/base/document_test.dart]: ah, ok.

**Pull request merged**